system/jlib/jutil.cpp

```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+        {
+            IPropertyTree &cat = iter->query();
+            const char *name = cat.queryName();
+            categories.emplace(name, std::unique_ptr<CVaultSet>(new CVaultSet(&cat)));
+        }
+    }
+    bool getCachedSecretFromVault(const char *category, const char *vaultId, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->getCachedSecretFromVault(vaultId, kind, content, secret, version);
+    }
+    bool requestSecretFromVault(const char *category, const char *vaultId, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->requestSecretFromVault(vaultId, kind, content, secret, version);
+    }
+
+    bool getCachedSecretByCategory(const char *category, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->getCachedSecret(kind, content, secret, version);
+    }
+    bool requestSecretByCategory(const char *category, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->requestSecret(kind, content, secret, version);
+    }
+};
+
+static Owned<CVaultManager> vaultManager;
+IVaultManager *ensureVaultManager()
+{
+    CriticalBlock block(secretCS);
+    if (!vaultManager)
+        vaultManager.setown(new CVaultManager());
+    return vaultManager;
+}
+
+CriticalSection secretCacheCS;
+static Owned<IPropertyTree> secretCache = createPTree();
+
+static IPropertyTree *getCachedLocalSecret(const char *name)
+{
+    if (isEmptyString(name))
+        return nullptr;
+    unsigned timeoutThreshold = msTick() - getSecretTimeout();
+    Owned<IPropertyTree> secret;
+    {
+        CriticalBlock block(secretCacheCS);
+        secret.setown(secretCache->getPropTree(name));
+        if (secret)
+        {
+            unsigned created = (unsigned) secret->getPropInt("@created");
+            if (created && (created < timeoutThreshold))
```

Comment:
nit picking: This could cause wrapping issues if created is a very large number - it could never time out.  You need to check if now - created > getSecretTimeout().